### ISSUE TITLE
TransferListener not firing periodically and not showing progress

### ISSUE THREAD

**Issue by jitendraramoliya** [awslabs/aws-sdk-android-samples#43]
I am using this code for upload video file. Video file is uploaded successfully but not getting progress periodically. its only one time show 0% progress. after then uploading video and getting "GC_FOR_ALLOC freed 1191K, 20% free 7353K/9108K, paused 4ms, total 4ms" message in log but not getting progress1 log and even not firing any method from transferlistener

TransferUtility transferUtilitys = new TransferUtility(sS3Client,
MainActivity.this);

        TransferObserver observer1 = transferUtilitys.upload(Constants.BUCKET_NAME, "test" + "/" + file.getName(),
                file, objectMetadata);

        observer1.setTransferListener(new TransferListener() {

            @Override
            public void onError(int id, Exception e) {
                Log.e(TAG, "Error during upload: " + id, e);
                Toast.makeText(MainActivity.this, "Error during upload: " + id, Toast.LENGTH_SHORT).show();

            }

            @Override
            public void onProgressChanged(int id, long bytesCurrent, long bytesTotal) {

                int progress = (int) ((double) bytesCurrent * 100 / bytesTotal);
                Log.e("", "progress1::" + progress);

            }

            @Override
            public void onStateChanged(int id, TransferState newState) {
                Log.e("", "state1::" + newState);
            }
        });

**Comment by fosterzhang** [comment]
I wasn't able to reproduce this. [S3TransferUtilitySample](https://github.com/awslabs/aws-sdk-android-samples/tree/master/S3TransferUtilitySample) shows the correct behavior of transfer listener. One thing suspicious is the empty tag in the log. If that's not the issue, please provide additional information.
- AWS Android SDK version
- Device/emulator info, such as Android platform version, device make, etc
- Complete code including the containing method and class.

**Comment by jitendraramoliya** [comment]
This code is working in new project but not showing progress in my project. It show progress only when I use this code. 

new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
                @Override
                public void run() {
                    int progress = (int) ((double) observer.getBytesTransferred() * 100 / observer.getBytesTotal());
                    Log.e("", "progress12::" + progress);
                }
            }, 20000);

but if process then greater than 20000 milisec then didn't get progress update. In uploade code in fragment and adapter. is it have problem with fragment and adapter. Could you check it with fragment and adapter. 

AWS version is 2.2.11
I am using Genemotion emulator and Samsung gallaxy s3 device.
device android version 4.3
I used same code that you  provided in UploadActivity in sample

**Comment by fosterzhang** [comment]
Thanks for reporting. I've identified the bug. The listener is garbage collected before the transfer finishes. See [aws-sdk-android #93](https://github.com/aws/aws-sdk-android/issues/93). Stay tuned for the fix.

**Comment by jitendraramoliya** [comment]
Ok. Let me know when you will get proper solution for it. I have found another issue while testing. I am pausing while uploading video. upload process is done around 49 percentage. and paused successfully. but when I resume that observer with same s3Id, I am getting resume from intially. I am getting process from 0 percentage and I am getting two listener display log at that time.. So my question is how can I get process resume where I paused last time.

**Comment by fosterzhang** [comment]
Please see my reply on the listener issue.

As for the upload progress, there is a limitation of S3 upload. Resuming an S3 upload requires multipart upload. Each part has to be completed before it can be resumed. Unfinished parts are discarded. The threshold to trigger multipart upload is 5mb. Here are some scenarios:
- A file is less than 5mb. The transfer will always restart from beginning.
- A 10mb file, split into two 5mb uploads. If both parts are paused at 2.5mb (50% total), when resumed, it has to start from beginning.
- A 10mb file, split into two 5mb uploads. If part 1 completes, and part 2 is at 2.5mb, when resumed, it starts from 5mb.

For download, S3 supports range get. So the transfer can pick up where it is left.

**Comment by jitendraramoliya** [comment]
Thanks for Multipart upload suggestion. my concern is what I am not getting Comlete status when some times listener not call onProgressChanged and onStateChanged. So I put updated database row when it upload will complete. but getting problem due not getting status for completion of video uploading.

**Comment by jitendraramoliya** [comment]
Hello fosterzhang 
I am still not getting progress in fragment and adapter. I did put code when get TransferState.COMPLETED in TransferListener. But I am not getting this method called. onProgressChanged method is calling only one time with progress 0 then never call. I am not getting call in onError and not found any error in upload. my file uploaded successfully on s3 server but not getting progress. Could you provide me proper solution and how to make long live TransferListener for gettting progresss untill get complete state.

**Comment by wpsboys** [comment]
I am getting the same issue. Transfer listener isn't called and the observer status isn't updated past waiting...

**Comment by fosterzhang** [comment]
@jbailey2010 and @wpsboys, an improvement is WIP. Stay tuned.

**Comment by uxight** [comment]
I am getting the same issue about transfer listener too.. is it not my fault? When do you think it will be fixed?

**Comment by fosterzhang** [comment]
This is improved in v2.2.13 release. Please update the SDK at https://aws.amazon.com/mobile/sdk/. Check out what is changed in the [change log](https://github.com/aws/aws-sdk-android/blob/master/CHANGELOG.md).

**Comment by YasirAmeen** [comment]
long _bytesCurrent = bytesCurrent;
                long _bytesTotal = bytesTotal;


                float percentage =  ((float)_bytesCurrent /(float)_bytesTotal * 100);
                Log.d("percentage","" +percentage);

                progressbar.setProgress(percentage);

**Comment by kvasukib** [comment]
Hi @YasirAmeen, What version of SDK are you using? We fixed this issue by making the notification threshold configurable in 2.6.6 (https://github.com/aws/aws-sdk-android/blob/master/CHANGELOG.md#release-266). Can you try in version >= 2.6.6 and see if that mitigates the problem? Thanks!
